### PR TITLE
Deprecate the low level DES functions.

### PR DESCRIPTION
**Pull request by paulidale**
Use of the low level DES functions has been informally discouraged for a
long time. We now formally deprecate them.  Applications should instead use the EVP APIs, e.g. `EVP_EncryptInit_ex`, `EVP_EncryptUpdate`, `EVP_EncryptFinal_ex`, and the equivalently named decrypt functions.

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by paulidale** [comment]
The outstanding major item is how to implement the `DES_crypt` and `DES_fcrypt` calls.

**Comment by paulidale** [comment]
The `DES_crypt` and `DES_fcrypt` functions are deprecated.  This in turn means that the `-crypt` option to the `passwd` application needs to be deprecated.

Since crypt was the default algorithm, this has been changed to the BSD MD5 algorithm.

**Comment by richsalz** [comment]
I think silently changing the default algorithm, which will depend on no-depracated configiuration, is a bad idea.  If not supporting deprecated, then make the default be an error.

File: apps/passwd.c [hunk 1]
```diff
@@ -841,9 +845,10 @@ static int do_passwd(int passed_salt, char **salt_p, char **salt_malloc_p,
     assert(strlen(passwd) <= pw_maxlen);
 
     /* now compute password hash */
-#ifndef OPENSSL_NO_DES
+#if !defined(OPENSSL_NO_DES) && !defined(OPENSSL_NO_DEPRECATED_3_0)
     if (mode == passwd_crypt)
         hash = DES_crypt(passwd, *salt_p);
+    */
```
Comment (levitte) [inline apps/passwd.c]: Spurious end if comment detected

File: apps/version.c [hunk 2]
```diff
@@ -126,9 +123,6 @@ int version_main(int argc, char **argv)
     if (options) {
         printf("options: ");
         printf(" %s", BN_options());
-#ifndef OPENSSL_NO_DES
-        printf(" %s", DES_options());
-#endif
```
Comment (levitte) [inline apps/version.c]: It's only deprecated, not removed

**Comment by paulidale** [comment]
CI failures are relevant.  A function can't be deprecated but used.

File: apps/version.c (hunk 2 above)
Comment (paulidale) [reply apps/version.c]: All the other deprecations removed the option printing....

**Comment by paulidale** [comment]
I've modified the crypt functions so they are callable as KDFs and not buried inside the passwd app.  The SHA based ones aren't working, they have some uglies to do with the salt and number of rounds which might be better refactored.

The alternative approach would be to bring the DES crypt code into the passwd app and leave the rest alone.

**Comment by paulidale** [comment]
A third option would be to take the character set conversions and unpacking out of the KDFs and put them into the passwd app.  I'm not sure why each algorithm did this itself...

**Comment by paulidale** [comment]
Moving this out of WIP.

**Comment by paulidale** [comment]
Note: the default alg for passwd has not been changed.

**Comment by paulidale** [comment]
Review?

**Comment by paulidale** [comment]
Merged to master, thanks for the feedback.

**Pull request closed without merging**